rumqttd/architecture.md

```diff
@@ -139,5 +139,94 @@ NOTE: Transition to ConnectionReady will schedule the connection for consumption
+
+   - Config Validation
+   - Configuring and creating the Router
+   - Starting the Router
+   - Maintaining a channel that communicates with the router
+
+   When you start the Broker, the following happens ->
+
+   1. The Router is started
+   2. (on a new thread) Creates the metrics server
+   3. (on a new thread) Starts all `mqttv4` servers
+   4. (on a new thread) Starts all `mqttv5` servers
+   5. (on a new thread) Starts the prometheus listener
+   6. (on the same thread) Starts the ConsoleLink to allow
+
+2. Router
+
+   Entity managed by the broker that controls the flow of data in `rumqttd`. The router is responsible for managing, authorizing, and scheduling the flow of data between components within and connected to `rumqttd`.
+
+   <!-- At any given time, the router's event loop must choose between handling incoming messages or sending outgoing messages. -->
+
+   The router is built on the actor pattern, where it works primarily by _reacting_ to events that happen. As and when an event is added to `ibufs`, the router reacts by checking its current state (for a certain connection or message) and dispatching the appropriate action. This can involve changing other in-memory data structures, or writing the message to `obufs` to make sure it's handled by the appropriate link.
+
+   An action of any sort on the router's end which involves communication usually involves the router adding a structure or packet to one of these buffers.
+
+   The router also shares Channels, which do not contain the events themselves, but rather contain notifications to events that might be added into the shared event buffers. This also has the added benefit of preventing lock contention between different futures/tasks, which would've happened if they had to keep checking the buffers
+
+   When you start the router, the following happens ->
+
+   1. Transmission Channels get created to and from the router
+   2. A `Graveyard` to store information to persist connections
+   3. A tracker for All the alerts and metrics that are necessary to be handled
+   4. A Map between subscriptions and devices gets created
+   5. The two communication buffers `ibufs` and `obufs` get initialized
+   6. Logs for message acknowledgements, subscriptions, and alerts get created
+   7. A Packet Cache is created
+   8. The core Scheduler Object itself is created
+
+   Once all necessary things have been initialized, the `spawn` method intializes the router's event loop on a separate thread.
+   This method returns a cloneable transmission channel that the Broker passes to all links so they can communicate with the router.
+
+3. Link
+
+   An Asynchronous and usually network-facing entity that Handles transmission of information to and from the Router. Information here can mean things including but not limited to
+
+   1. Messages incoming from devices (publish)
+   2. Messages outgoing to devices (subscribe)
```

Comment:
there can be other things like connect, pings etc that can't be categorized as publish / subscribe ig

Reply:
I'll include these, too.